crates/subspace-networking/src/node_runner.rs

```diff
@@ -214,18 +220,36 @@ where
     }
 
     // Handle DSN online status signaling
-    fn signal_online_status(&mut self) {
+    fn signal_online_status(&mut self) -> bool {
         let current_online_status = self.swarm.connected_peers().next().is_some();
         let previous_online_status = *self.online_status_observer_tx.borrow();
 
         if previous_online_status != current_online_status {
-            if let Err(err) = self.online_status_observer_tx.send(current_online_status) {
-                error!("DSN connection observer channel failed: {err}")
+            let online_checks_number = self
+                .successful_online_status_check_number
+                .load(Ordering::SeqCst);
+
+            // We send 'offline' status immediately or wait for several successful checks before
+            // sending 'online' status to counter connection establishing/closing events on
+            // hitting the connection limits.
+            if !current_online_status
+                || online_checks_number >= STATUS_CHECK_NUMBER_FOR_RESTORING_OPERATIONS
+            {
+                if let Err(err) = self.online_status_observer_tx.send(current_online_status) {
+                    error!("DSN connection observer channel failed: {err}")
+                }
             }
         }
+
+        current_online_status
     }
 
     async fn handle_peer_dialing(&mut self) {
+        if self.signal_online_status() {
+            self.successful_online_status_check_number
+                .fetch_add(1, Ordering::SeqCst);
+        }
```

Comment:
I strongly dislike this. As mentioned many times before, doing more and more work in time intervals is a path towards DoS. If you're offline, you know about that immediately and the correct way to check it is in the even handling reactively, the way it was done before.

Reply:
I'm not sure I follow 100%. The offline message (connection closed) is acknowledged immediately, as you suggested. On the other hand, "connection established" is a weaker signal. We can't rely on it to send the "online message"; we need to check several times that the "connection established" didn't hit the connection limits. Could you describe more about what you have in mind?